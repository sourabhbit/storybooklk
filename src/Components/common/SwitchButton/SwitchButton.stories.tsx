import { ComponentMeta, ComponentStory } from '@storybook/react'
import { sizesENUM } from './SwitchButton.types'
import SwitchButton from './index'

export default {
	title: 'Components/common/SwitchButton',
	component: SwitchButton,
	argTypes: {
		name: {
			table: {
				category: 'General',
			},
		},
		id: {
			table: {
				category: 'General',
			},
		},
		disabled: {
			table: {
				category: 'Status',
			},
		},
		onClick: {
			table: {
				category: 'Events',
			},
		},
	},
} as ComponentMeta<typeof SwitchButton>

const Template: ComponentStory<typeof SwitchButton> = (args) => (
	<SwitchButton {...args} />
)

export const Overview = Template.bind({})
Overview.args = {
	name: 'button',
	id: 'button-id',
	leftText: 'VIEW FRAMES',
	rightText: 'VIEW 3D',
	currentStatus: false,
	size: sizesENUM.small,
}
