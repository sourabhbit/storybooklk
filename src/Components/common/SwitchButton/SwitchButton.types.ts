export interface SwitchButtonTypes {
	name: string
	leftText?: string
	rightText?: string
	currentStatus?: boolean
	size?: sizesENUM
	id: string
	onClick?: () => void
}

export enum sizesENUM {
	small = 'small',
	medium = 'medium',
	large = 'large',
}
