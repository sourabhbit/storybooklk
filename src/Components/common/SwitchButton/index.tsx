import { useState, useEffect } from 'react'
import {
	SwitchButtonWrapper,
	SwitchWrapper,
	CmnToggleInputWrapper,
	CmnToggleLabelWrapper,
	LeftSpanWrapper,
	RightSpanWrapper,
} from './styles'
import { SwitchButtonTypes, sizesENUM } from './SwitchButton.types'

/**
 * Default Switchbutton Component
 */
const Switchbutton = ({
	id,
	name,
	currentStatus,
	leftText,
	rightText,
	onClick = () => null,
	size = sizesENUM.medium,
}: SwitchButtonTypes) => {
	const [left, setLeft] = useState(currentStatus)
	const [checked, setChecked] = useState(false)

	const handleToggle = () => {
		setLeft(!left)
		onClick()
	}
	useEffect(() => {
		if (left) setChecked(false)
		else setChecked(true)
	}, [left])
	return (
		<SwitchButtonWrapper data-testid="switchButton">
			{leftText && (
				<LeftSpanWrapper
					data-testid="switch-leftText"
					styleSize={size}
					checked={checked}
				>
					{leftText}
				</LeftSpanWrapper>
			)}
			<SwitchWrapper>
				<CmnToggleInputWrapper
					type="checkbox"
					id={id}
					name={name}
					data-testid="switch-input"
				/>
				<CmnToggleLabelWrapper
					checked={checked}
					onClick={() => handleToggle()}
					data-testid="switch-label"
				/>
			</SwitchWrapper>
			{rightText && (
				<RightSpanWrapper
					data-testid="switch-rightText"
					styleSize={size}
					checked={checked}
				>
					{rightText}
				</RightSpanWrapper>
			)}
		</SwitchButtonWrapper>
	)
}

export default Switchbutton
