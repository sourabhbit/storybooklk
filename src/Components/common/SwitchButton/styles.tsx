import styled from 'styled-components'
import { tokens } from '../../../Styles/tokens'
import { sizesENUM } from './SwitchButton.types'

export const SwitchButtonWrapper = styled.div`
	display: flex;
	align-items: center;
	font-size: 1.2vw;
	line-height: 1.4;
	letter-spacing: 1.5px;
`

export const SwitchWrapper = styled.div`
	padding-right: var(--pd-20);
	padding-left: var(--pd-20);
`

export const CmnToggleInputWrapper = styled.input`
	position: absolute;
	visibility: hidden;
	padding: 2px;
`

export const LeftSpanWrapper = styled.span<{
	checked: boolean
	styleSize: sizesENUM
}>`
	color: ${(props) =>
		!props.checked ? 'var(--blashish-gray)' : 'var(--black)'};
	font-size: ${(props) => tokens.fontSize[props.styleSize]};
`

export const RightSpanWrapper = styled.span<{
	checked: boolean
	styleSize: sizesENUM
}>`
	color: ${(props) =>
		!props.checked ? 'var(--black)' : 'var(--blashish-gray)'};
	font-size: ${(props) => tokens.fontSize[props.styleSize]};
`

export const CmnToggleLabelWrapper = styled.label<{
	checked: boolean
}>`
	display: block;
	position: relative;
	cursor: pointer;
	outline: none;
	user-select: none;
	padding: 2px;
	width: 70px;
	height: 25px;
	background-color: #dddddd;
	border-radius: 60px;
	&:before {
		display: block;
		position: absolute;
		top: 1px;
		left: 1px;
		bottom: 1px;
		content: '';
		right: 1px;
		background-color: ${(props) =>
			!props.checked ? 'var(--blashish-gray)' : 'var(--whitish-gray)'};
		border-radius: 60px;
		transition: background 0.4s;
		&:checked {
			background-color: var(--blashish-gray);
		}
	}
	&:after {
		display: block;
		position: absolute;
		top: 1px;
		left: ${(props) => (!props.checked ? '1px' : 'inherit')};
		right: ${(props) => (props.checked ? '1px' : 'inherit')};
		bottom: 1px;
		content: '';
		width: 24px;
		background-color: var(--white);
		border-radius: 100%;
		box-shadow: 0 2px 5px rgba(0, 0, 0, 0.3);
		transition: margin 0.4s;
		&:checked {
			margin-left: 44px;
		}
	}
`
