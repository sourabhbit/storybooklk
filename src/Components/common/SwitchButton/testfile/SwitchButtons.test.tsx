import '@testing-library/jest-dom'
import { render, screen, fireEvent } from '@testing-library/react'
import { sizesENUM } from '../SwitchButton.types'
import Switchbutton from '../index'

const props = {
	name: 'switch button',
	id: 'button-id',
	leftText: 'VIEW FRAMES',
	rightText: 'VIEW 3D',
	currentStatus: false,
	size: sizesENUM.small,
	onClick: jest.fn(),
}

describe('<Switchbutton />', () => {
	let _asFragment: any = null
	let _rerender: any = null

	beforeEach(() => {
		const { asFragment, rerender } = render(<Switchbutton {...props} />)
		_asFragment = asFragment
		_rerender = rerender
	})

	afterAll(() => {
		_asFragment = null
	})

	it('Should render component', () => {
		const container = screen.getByTestId('switchButton')
		expect(container).toBeInTheDocument()
	})

	it('text should render', () => {
		const leftText = screen.getByTestId('switch-leftText')
		const rightText = screen.getByTestId('switch-rightText')
		expect(leftText).toBeInTheDocument()
		expect(rightText).toBeInTheDocument()
	})

	it('text should not render', () => {
		_rerender(<Switchbutton {...props} leftText={''} rightText={''} />)
		const leftText = screen.queryByTestId('switch-leftText')
		const rightText = screen.queryByTestId('switch-rightText')
		expect(leftText).not.toBeInTheDocument()
		expect(rightText).not.toBeInTheDocument()
	})

	it('onclick function should run', () => {
		fireEvent.click(screen.getByTestId('switch-label'))
		expect(props.onClick).toHaveBeenCalledTimes(1)
	})

	it('Should take a snapshot', () => {
		expect(_asFragment()).toMatchSnapshot()
	})
})
