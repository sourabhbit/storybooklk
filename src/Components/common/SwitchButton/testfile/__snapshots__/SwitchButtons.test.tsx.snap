// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switchbutton /> Should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL gzLLZV"
    data-testid="switchButton"
  >
    <span
      class="sc-hKwDye jtIkzO"
      data-testid="switch-leftText"
    >
      VIEW FRAMES
    </span>
    <div
      class="sc-gsDKAQ HmPLt"
    >
      <input
        class="sc-dkPtRN eVNsQd"
        data-testid="switch-input"
        id="button-id"
        name="switch button"
        type="checkbox"
      />
      <label
        class="sc-jRQBWg eojDAa"
        data-testid="switch-label"
      />
    </div>
    <span
      class="sc-eCImPb dxlSiH"
      data-testid="switch-rightText"
    >
      VIEW 3D
    </span>
  </div>
</DocumentFragment>
`;
